Migrate System.Runtime.CompilerServices.Unsafe

carlossanlop opened the pull request:
Migrated from 6.0.

The [System.Runtime.CompilerServices.Unsafe nuget package](https://www.nuget.org/packages/System.Runtime.CompilerServices.Unsafe) has this information:

- Supports .NET Standard 2.0.
- Has public types in .NET Framework 4.6.2 and .NET Standard 2.0.

- Has no ref assemblies.
- Has no PNSE.

When consuming this nuget package in a console app, then printing `typeof(System.Runtime.CompilerServices.Unsafe).Assembly.CodeBase` I confirmed that:

- When targeting .NET Framework 4.6.2, this assembly is used: `bin/Debug/net462/System.Runtime.CompilerServices.Unsafe.DLL`

- When targeting .NET 6.0, this assembly is used: `file:///C:/Program Files/dotnet/shared/Microsoft.NETCore.App/6.0.25/System.Runtime.CompilerServices.Unsafe.dll`

carlossanlop commented:
@ViktorHofer brought this up today - 6.0 is still in support, and that's the source branch for this package. I think we should just close this PR and migrate the package when 6.0 goes out of support. Any objections @ericstj?

ericstj commented:
Suppose 6.0 goes out of support, then the very next day we have a need to service Unsafe urgently.  I wouldn't want to keep building the 6.0 branch.  I'd prefer we have servicing ready for Unsafe from this codebase.  We don't need to get this done right away but I want it to be part of the first wave of packages so that folks can move over before we need to urgently service it.

carlossanlop commented:
Okay. Makes sense.

I can open an issue so that when .NET 6.0 goes out of support, we double check that we have already cherry-picked any and all commits that were added to the assembly in 6.0 after the migration (very low chance, but just in case).

ViktorHofer commented:
If you plan to merge this PR in before 6.0 is out-of-support, consider not referencing this library from other projects in this repo as this could easily result in a wrong assembly version of S.R.CS.U referenced on .NET Framework.

IMHO it would better if we prepare this PR but not merge it in until November to avoid potential confusion and not go against the migration criteria from https://github.com/dotnet/maintenance-packages/issues/42#issue-2096826599:

> The component is not being built out of main or a branch that is still serviced (.NET 6, .NET 7 and .NET 8).

ericstj commented:
We actually want to ship the new version and get folks to move to it - think of this repo like a new LTS for all these packages.  It's totally OK for that version to exist in parallel with the 6.0.0 branch.  It should be OK to ship from it in parallel - just like we do with packages coming out of the 8.0 branches and 6.0 branches.

The 6.0 branch living on is not really much different from the 3.1/2.1 branches still being around if we need to do *urgent* servicing with a minimal diff.

> If you plan to merge this PR in before 6.0 is out-of-support, consider not referencing this library from other projects in this repo as this could easily result in a wrong assembly version of S.R.CS.U referenced.

Where would that happen?  We'd always be referencing the one from this repo - which is what we want.

ViktorHofer commented:
> Where would that happen? We'd always be referencing the one from this repo - which is what we want.

I'm talking about the case when S.R.CS.U services in runtime release/6.0 and updates its assembly version and packages from this repository that depend on it publish. They would point to the un-serviced assembly version of Unsafe in this repo.

> The 6.0 branch living on is not really much different from the 3.1/2.1 branches still being around if we need to do urgent servicing with a minimal diff.

If that's the direction, fine but please update the migration criteria to reflect that. Right now it explicitly mentions that packages from still serviced branches aren't a candidate.

ericstj commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj:
```diff
@@ -0,0 +1,97 @@
+﻿<Project Sdk="Microsoft.NET.Sdk.IL">
+  <PropertyGroup>
+    <TargetFrameworks>net6.0;netcoreapp3.1;netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```
Probably need to remove these frameworks and perhaps add a placeholder.

ericstj commented:
> I'm talking about the case when S.R.CS.U services in runtime release/6.0 and updates its assembly version and packages from this repository that depend on it publish. They would point to the un-serviced assembly version of Unsafe in this repo.

I see. We'd sim-ship a new release from this repo.  Much the same as we would sim-ship a release from 8.0 when 6.0 was serviced.

When we started this project, we said that the packages produced from here are different enough from the packages in previous releases that we wanted to give folks notice to move to the new version before we make using it the *only* way to get a critical update - just like we do for new LTS releases.  We said we wanted to publish these and ask folks to move to them within a certain time window.  During that time window - if a critical fix is needed - we still need to ship it from both here and the old location, so that folks had a pathway with minimal changes to get a critical fix.

Now it's completely possible that we treat Unsafe differently - it's possible it has fewer changes from its previous version - enough that we might just consider it a continuation of 6.0.0.  To do that we shouldn't remove anything from it and do some extra testing to make sure it's effectively the same as 6.0.  If that's the case then we could probably delay this PR and then version it as if it still was coming from the 6.0 branch.

ViktorHofer commented on src/System.Runtime.CompilerServices.Unsafe/System.Runtime.CompilerServices.Unsafe.sln:
```diff
@@ -0,0 +1,35 @@
+﻿Microsoft Visual Studio Solution File, Format Version 12.00
```
This file should be deleted. We don't add slns per library in this repo.

carlossanlop replied:
Yep. I'm aware. I just haven't cleaned it up properly after migration.

ViktorHofer commented:
I'm not sure you understand what my proposal is. My recommendation is the following:
1. Pause this PR for eight months until .NET 6 is EOL (November 12th, 2024). Don't merge it in to avoid unnecessarily sim-shipping this package in the event of a servicing request.
2. Update this PR so that the library doesn't target TFMs that will be out-of-support after November 12th. This means removing all .NETCoreApp TFMs and adding a placeholder TFM for net7.0 as that's the runtime that has Unsafe inbox.
3. Ship a new version of this library on November 13th or shortly afterwards with a minor version bump.
4. From then on, use this codebase to service the Unsafe nuget package.

This should give customers enough time to realize the TFM change when we ship the new minor version until a servicing request comes in.

carlossanlop commented on global.json:
```diff
@@ -14,6 +14,7 @@
     "Microsoft.DotNet.Arcade.Sdk": "8.0.0-beta.24165.4",
-    "Microsoft.Build.Traversal": "3.4.0"
+    "Microsoft.Build.Traversal": "3.4.0",
+    "Microsoft.NET.Sdk.IL": "8.0.0"
```
From https://github.com/dotnet/runtime/issues/11411

carlossanlop commented:
The tests are passing, packing succeeds, and no api compat failures show up. Which makes me suspect something's wrong 😄. Can I get a review, please?

I inspected the generated nupkg when packing and I can see that the `libs/` folder contains a `net462/` folder and a `netstandard2.0/`, each with a DLL.

At the root level, there is another folder called `buildTransitive` which contains two subfolders: `net462/` and, strangely, a `net461/` folder. The `net462/` folder contains a `_._` placeholder file. But the `net461/` folder contains a `System.Runtime.CompilerServices.Unsafe.targets` file with the following contents:

```xml
<Project InitialTargets="NETStandardCompatError_System_Runtime_CompilerServices_Unsafe_net462">
  <Target Name="NETStandardCompatError_System_Runtime_CompilerServices_Unsafe_net462"
          Condition="'$(SuppressTfmSupportBuildWarnings)' == ''">
    <Warning Text="System.Runtime.CompilerServices.Unsafe 1.0.0-dev.24175.1 doesn't support $(TargetFramework) and has not been tested with it. Consider upgrading your TargetFramework to net462 or later. You may also set &lt;SuppressTfmSupportBuildWarnings&gt;true&lt;/SuppressTfmSupportBuildWarnings&gt; in the project file to ignore this warning and attempt to run in this unsupported configuration at your own risk." />
  </Target>
</Project>
``` 

How did this get generated?

carlossanlop commented:
Update: I am seeing this same `buildTransitive/net461/` folder in other nupkgs. I am leaning more to the option that this is expected, but I'm still curious to learn how it gets added.

ViktorHofer commented:
Needs a placeholder file for net7.0+.
The buildTransitive targets are added in packaging.targets and are expected.

carlossanlop commented:
> Needs a placeholder file for net7.0+.

How did you figure this out?

ViktorHofer commented:
I added that infrastructure as part of https://github.com/dotnet/maintenance-packages/commit/c4c0f96d6844d2d7886b2c9320a5be5913190ffd

carlossanlop commented:
> We moved S.R.CompilerServices.Unsafe inbox during .NET 7 and stopped producing a package. As it's inbox, the netstandard2.0 asset shouldn't be selected when targeting net7.0 or greater. A placeholder file achieves that.

Ah, okay. The regular way of figuring out as we did with other packages with the same problem. I wasn't sure about Unsafe as this package is weird with all the System.Private.CoreLib dependencies.

carlossanlop commented:
To unblock this PR, I included the fix from https://github.com/dotnet/maintenance-packages/pull/85 , but I want to merge that change as a separate commit.

carlossanlop commented:
I'm seeing a new failure that I can't repro locally: https://dev.azure.com/dnceng-public/public/_build/results?buildId=625996&view=logs&j=a751307a-bba0-598c-20e2-725c5e776a43&t=bd0b5859-55f0-5511-bafc-d48b64174767

The failing target is this one: D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\8.0.0-beta.24179.4\tools\SymStore.targets

```xml
  <Target Name="_DeployPortableSymbolsToSymStore"
          DependsOnTargets="_DeployToSymStoreSetProperties"
          Inputs="$(TargetPath);$(_TargetPdbPath)"
          Outputs="$(_SymStorePdbPath)"
          Condition="'$(PublishWindowsPdb)' == 'true' and $(_BuildsPortablePdb)">

    <PropertyGroup>
      <_PdbConverterPath>$(NuGetPackageRoot)microsoft.diasymreader.pdb2pdb\$(MicrosoftDiaSymReaderPdb2PdbVersion)\tools\Pdb2Pdb.exe</_PdbConverterPath>
      <_PdbConverterCommandLineArgs>"$(TargetPath)" /out "$(_SymStorePdbPath)" /srcsvrvar SRC_INDEX=public</_PdbConverterCommandLineArgs>
    </PropertyGroup>

    <Error Text="Attempt to publish Portable PDB to SymStore without conversion: UsingToolPdbConverter property is false in project $(MSBuildProjectName)"
           Condition="'$(UsingToolPdbConverter)' != 'true'"
           ContinueOnError="false"/>
```

```
Task "Error" skipped, due to false condition; ('$(UsingToolPdbConverter)' != 'true') was evaluated as ('true' != 'true').
```

The binlog says the `UsingToolPdbConverter` property is set to true everywhere. I am unsure what this tool is supposed to achieve. Do we need this target to get processed, @ericstj?

Artifacts: https://dev.azure.com/dnceng-public/public/_build/results?buildId=625996&view=artifacts&pathAsName=false&type=publishedArtifacts

ericstj commented:
I think there's likely a step that's not being run in this ILProj.  Maybe you can check the binlog to see if some tool is being run to transform the PDBs produced by CSProj's and then see how that's scheduled to see if we can make it work for the ILProj.  You should be able to repro locally if you build the same way that the CI does.

ericstj commented:
Oh, I looked closer - you weren't hitting the error that you copied - you're just hitting an error because the PDB doesn't exist.  Double check that you're producing a PDB from the ilproj.  I think there might be some settings you need to make to get that to work correctly.

carlossanlop commented:
Ok, I was able to locally repro the same error as in the CI.

The symstore stuff is located in `C:\.tools\.nuget\packages\microsoft.dotnet.arcade.sdk\8.0.0-beta.24179.4\tools\SymStore.targets`.

I first confirmed that the `<MakeDir>` task inside the `_DeployPortableSymbolsToSymStore` target is creating the `$(_SymStoreOutputDir)` correctly for the Unsafe project.

I can also confirm that the DLL that is supposed to contain the embedded PDB does indeed exist in the expected location: `C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Runtime.CompilerServices.Unsafe\Release\net462\System.Runtime.CompilerServices.Unsafe.dll`

The help output of the Pdb2Pdb.exe tool says the following:

```
C:\.tools\.nuget\packages\microsoft.diasymreader.pdb2pdb\1.1.0-beta2-19521-03\tools\Pdb2Pdb.exe
Usage: Pdb2Pdb <dll/exe path> [/pdb <path>] [/out <path>] [/extract]

/pdb <path>                Path to the PDB to convert. If not specified explicitly,
                           the PDB referenced by or embedded in the DLL/EXE is used.
/out <path>                Output PDB path.
/extract                   Extract PDB embedded in the DLL/EXE.
/sourcelink                When converting to Windows PDB do not convert SourceLink to srcsrv format.
                           By default both SourceLink and srcsrv are included in the converted PDB.
/nowarn *                  Suppress all warnings. All warnings are reported by default.
/nowarn <n1,n2,...>        Suppress specified warning.
/srcsvrvar <name>=<value>  Add specified variable to srcsvr stream. Only applicable
                           when converting to Windows PDB and /sourcelink is not
                           specified.

/extract and /pdb are mutually exclusive.


Missing dll/exe path.
```

Note that we're not using the `/pdb` argument, we're just directly passing the DLL as the `$(TargetPath)`, so I'd like to think that `/pdb` is assumed.

Here's the command that is executed according to my binlog:
```
$ C:\.tools\.nuget\packages\microsoft.diasymreader.pdb2pdb\1.1.0-beta2-19521-03\tools\Pdb2Pdb.exe "C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Runtime.CompilerServices.Unsafe\Release\net462\System.Runtime.CompilerServices.Unsafe.dll" /out C:\Users\calope\source\repos\maintenance-packages\artifacts\SymStore\Release\System.Runtime.CompilerServices.Unsafe\net462\System.Runtime.CompilerServices.Unsafe.pdb /srcsvrvar SRC_INDEX=public
```

And this is the output:
```
File 'C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Runtime.CompilerServices.Unsafe\Release\net462\System.Runtime.CompilerServices.Unsafe.pdb' not found or doesn't match 'C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Runtime.CompilerServices.Unsafe\Release\net462\System.Runtime.CompilerServices.Unsafe.dll'.
```

@ericstj If we were assuming that the PDB was embedded in the DLL, where can I check that such content is indeed located in the DLL?

ericstj commented:
I wonder how it was working in 6.0?  Maybe could examine the build there to see what expected behavior should be.  You could determine 1) did we produce a PDB there?  2) did we run this target there? etc etc.

carlossanlop commented:
Thanks @ericstj for your help figuring out the root cause. I pushed the commits. 🤞 Hope that gives us a green CI.

carlossanlop commented:
The CI is green, @ericstj. But I still need to squash all my commits, which takes some time since now we have merges from main inserted in the middle.

ericstj commented on Directory.Build.targets:
```diff
@@ -38,4 +38,14 @@
 
+  <!-- This workaround prevents the execution of the RunTests target even when SkipTests is set to true.
+  It can be removed after this bug is fixed: https://github.com/dotnet/arcade/pull/14665 -->
+  <Target Name="_SetSkipTests" AfterTargets="_OuterGetTestsToRun">
```
This should be removed.

ericstj commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj:
```diff
@@ -0,0 +1,94 @@
+          BeforeTargets="DisassembleIlasmResourceFile"
+          Condition="'@(ResolvedMatchingContract)' != ''">
+    <ItemGroup>
+      <ILResourceReference Include="@(ResolvedMatchingContract)" />
```
Is this working correctly?  I noticed that the commandline to ILASM didn't have resource.  In this repo we don't build a contract assembly, so you might want to use some other CSProj to produce a binary that can be used for grabbing native version resources.

carlossanlop replied:
I don't really understand this. Can you explain what you mean by using another csproj?

ericstj replied:
Ilasm doesn't have a way for defining PE version resources in IL.  Instead what you need to do is pass in a resource object file.

Rather than having a scheme for authoring native resources as `.rc` and flowing all the right branding info into that ILProj uses a trick.  It produces resources for the ILProj by extracting them from some other assembly built with CSC.  https://github.com/dotnet/runtime/blob/57f870f909dbfad35142e5aaa6e681464de4f439/src/coreclr/.nuget/Microsoft.NET.Sdk.IL/Microsoft.NET.Sdk.IL.targets.template#L71-L111

So you just need an assembly that's versioned appropriately and that needs to be passed in as the `ILResourceReference` item.  You'll be able to see if you've gotten it correct by examining the version properties of the file (for instance in Windows Explorer).

Setting the native version resources is important because this is used by installers for versioning.

Related: https://github.com/dotnet/runtime/issues/11412

carlossanlop replied:
> So you just need an assembly that's versioned appropriately 

I added the missing versioning properties in the ilproj. But if I look at the properties of the 

![image](https://github.com/user-attachments/assets/d081830b-8a33-43c3-8c45-685d57a4a564)

According to the binlog, that target is not getting evaluated due to the condition not being met:

> Target "SetILResourceReference" skipped, due to false condition; ('@(ResolvedMatchingContract)' != '') was evaluated as ('' != '').

I'm now trying to understand how or where we gave a value to `ResolvedMatchingContract` in release/6.0.

ericstj replied:
That's coming from the reference assembly resolution.  We won't have a reference assemblies here, so we can't use that to feed into the versioning.  Instead we just need some other CSProj to feed in for version information.  We could create a dummy one (probably the cleanest) or use an existing one (hacky, but doesn't add projects).  In either case we'd need to be careful to not include that reference in the package (as dependency or file).

ViktorHofer commented on Build.proj:
```diff
@@ -1,7 +1,8 @@
 
   <ItemGroup>
-    <ProjectReference Include="src/**/*.csproj" />
+    <ProjectReference Include="src/*/src/*.csproj" />
+    <ProjectReference Include="src/*/src/*.ilproj" />
```
```suggestion
    <ProjectReference Include="src/*/src/*.*proj" />

```

ViktorHofer commented on src/System.Runtime.CompilerServices.Unsafe/tests/System.Runtime.CompilerServices.Unsafe.Tests.csproj:
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(NetFrameworkMinimum)</TargetFrameworks>
```
```suggestion
    <TargetFramework>$(NetFrameworkMinimum)</TargetFramework>
```

ViktorHofer commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj:
```diff
@@ -0,0 +1,94 @@
+
+  <PropertyGroup Condition="'$(TargetFramework)' == '$(NetFrameworkMinimum)'">
+    <CoreAssembly>mscorlib</CoreAssembly>
+    <ExtraMacros>#define $(TargetFramework)</ExtraMacros>
```
This macro isn't used in the .il file anywhere and can be removed.

ViktorHofer commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj:
```diff
@@ -0,0 +1,94 @@
+<Project Sdk="Microsoft.NET.Sdk.IL">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(NetFrameworkMinimum);netstandard2.0;net7.0</TargetFrameworks>
```
I wonder what would happen if someone references this library on the .NET 6 out-of-support runtime. Would it work or blow up epically?

carlossanlop replied:
How should I test this?

ViktorHofer replied:
Test the package with a .NET 6 runtime?

carlossanlop replied:
All NETCoreApp versions are marked as placeholders:
```xml
<IsPlaceholderTargetFramework Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">true</IsPlaceholderTargetFramework>
```
Let me see what happens.

ViktorHofer replied:
> All NETCoreApp versions are marked as placeholders:

I'm not sure what you mean by that. We only target one .NETCoreApp version in this project: net7.0.

carlossanlop replied:
Ok I see, the placeholder does not apply here even though the condition is stated as I pasted above. This means that if we reference this library via .NET 6.0, we would go through .NET Standard 2.0: https://learn.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-2-0

ViktorHofer replied:
Correct

carlossanlop replied:
@ericstj I tested consuming this new package using net6.0 and posted the result here: https://github.com/dotnet/maintenance-packages/pull/59#issuecomment-2256761751 

Since net6.0 is not yet out of support, I am unsure what to expect. I suppose we will end up trying to consume the netstandard2.0 version of the package that comes out from here, but then what?

ViktorHofer commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj:
```diff
@@ -0,0 +1,94 @@
+
+  <PropertyGroup Condition="'$(TargetFramework)' == '$(NetFrameworkMinimum)'">
+    <CoreAssembly>mscorlib</CoreAssembly>
+    <ExtraMacros>#define $(TargetFramework)</ExtraMacros>
```
I'm struggling to find any place that uses this define.

carlossanlop replied:
The only macro I found in the *.il file is `netcoreapp`

carlossanlop replied:
Were you looking in other places also?

carlossanlop replied:
Actually, if I remove it, the *.il fails to build. I need to keep it. This is the error:

```

C:\Users\calope\source\repos\maintenance-packages\artifacts\obj\System.Runtime.CompilerServices.Unsafe\Debug\net462\version\version.h(9): error : syntax error at token '{' in: { [C:\Users\calope\source\re 
pos\maintenance-packages\src\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj::TargetFramework=net462]
C:\.tools\.nuget\packages\microsoft.net.sdk.il\8.0.0\targets\Microsoft.NET.Sdk.IL.targets(136,5): error MSB3073: The command ""C:\.tools\.nuget\packages\\runtime.win-x64.microsoft.netcore.ilasm\6.0.0\runt 
imes\win-x64\native/ilasm" -QUIET -NOLOGO -DLL  -DEBUG=IMPL -INCLUDE="C:\Users\calope\source\repos\maintenance-packages\artifacts\obj\System.Runtime.CompilerServices.Unsafe\Debug\net462\version" -OUTPUT=" 
C:\Users\calope\source\repos\maintenance-packages\artifacts\obj\System.Runtime.CompilerServices.Unsafe\Debug\net462\System.Runtime.CompilerServices.Unsafe.dll" -KEY="C:\.tools\.nuget\packages\microsoft.do 
tnet.arcade.sdk\8.0.0-beta.24360.5\tools\snk/MSFT.snk" System.Runtime.CompilerServices.Unsafe.il" exited with code 1. [C:\Users\calope\source\repos\maintenance-packages\src\System.Runtime.CompilerServices 
.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj::TargetFramework=net462]
```

That version.h file is generated with the define like this:
```c
#define CORE_ASSEMBLY "mscorlib"
#define ASSEMBLY_VERSION "1:0:0:0"
#define CORE_ASSEMBLY_VERSION "4:0:0:0"
#define FILE_VERSION "{string('1.0.24.36201')}"
#define INFORMATIONAL_VERSION "{string('1.0.0-dev.24362.1')}"
#define net462
// Metadata version: v4.0.30319
.assembly extern CORE_ASSEMBLY
{
  .publickeytoken = (B7 7A 5C 56 19 34 E0 89 )
  .ver CORE_ASSEMBLY_VERSION
}
```

ViktorHofer commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il:
```diff
@@ -0,0 +1,648 @@
+// https://github.com/dotnet/runtime/issues/8730
+// and
+// https://github.com/dotnet/coreclr/pull/11218
+#ifdef netcoreapp
+        .maxstack 1
+        ldarg.0
+        ret
+#else
```
We don't compile for `netcoreapp` anymore. Should these branches be deleted?

carlossanlop replied:
Are you saying that everything after (and including) .NET 5 is not considered .NETCoreApp, but only NET? I thought NET was a subgroup of NETCoreApp.

carlossanlop replied:
Removed

carlossanlop commented on src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj:
```diff
@@ -0,0 +1,94 @@
+<Project Sdk="Microsoft.NET.Sdk.IL">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(NetFrameworkMinimum);netstandard2.0;net7.0</TargetFrameworks>
```
Should I keep net7.0 now that it's out of support?

ViktorHofer replied:
`net7.0` doesn't get compiled, it's just a "placeholder" TFM to create a placeholder file. So it should be kept.

carlossanlop replied:
The placeholder condition applies to all .NET Core App versions:
```xml
<IsPlaceholderTargetFramework Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">true</IsPlaceholderTargetFramework>
```

carlossanlop replied:
But as you said in another comment, we don't compile against netcoreapp, so I removed the preprocessor branches that applied to it.

I mainly ask since you left a question if we would have a bad failure when consuming this in .NET 6.0.

ViktorHofer replied:
> The placeholder condition applies to all .NET Core App versions:

Yes but it was net7.0 when we moved the API inbox so the placeholder needs to be under the "lib/net7.0" folder in the package.

carlossanlop commented:
I'm coming back to this.  I still need to address the above feedback.

But first: I noticed I did not add AssemblyVersion, VersionPrefix, PackageValidationBaselineVersion. 😖 No wonder the build was passing.

Let me refresh my brain, and I'll also share my thoughts out loud. This is what I know:

- Nuget: https://www.nuget.org/packages/System.Runtime.CompilerServices.Unsafe
- Supports .NET Standard 2.0.
- Has public types in .NET Framework 4.6.2 and .NET Standard 2.0.
- Has no ref assemblies.
- Has no PNSE.
- When consuming this nuget package in a console app, then printing `typeof(System.Runtime.CompilerServices.Unsafe).Assembly.CodeBase` I confirmed that:
- When targeting .NET Framework 4.6.2, this assembly is used: `bin/Debug/net462/System.Runtime.CompilerServices.Unsafe.DLL`
- When targeting .NET 6.0, this assembly is used: `file:///C:/Program Files/dotnet/shared/Microsoft.NETCore.App/6.0.25/System.Runtime.CompilerServices.Unsafe.dll`
- The packageIndex.json file in 3.1 says the following for this package: https://github.com/dotnet/corefx/blob/b58ef0a19eca6269799664d6288be14b7f0cef3e/pkg/Microsoft.Private.PackageBaseline/packageIndex.json#L4246-L4272

```json
    "System.Runtime.CompilerServices.Unsafe": {
      "StableVersions": [
        "4.0.0",
        "4.3.0",
        "4.4.0",
        "4.5.0",
        "4.5.1",
        "4.5.2",
        "4.6.0",
        "4.7.0",
        "4.7.1"
      ],
      "BaselineVersion": "4.7.1",
      "InboxOn": {
        "netcoreapp3.0": "4.0.5.0",
        "netcoreapp3.1": "4.0.6.0",
        "uap10.0.16300": "4.0.6.0"
      },
      "AssemblyVersionInPackageVersion": {
        "4.0.3.0": "4.4.0",
        "4.0.4.0": "4.5.0",
        "4.0.4.1": "4.5.2",
        "4.0.5.0": "4.6.0",
        "4.0.6.0": "4.7.0",
        "4.0.7.0": "4.7.1"
      }
    },
```

Viktor mentioned above that this could blow up if we use it in net6.0. I think he wants me to verify what would happen because we still published a version of Unsafe in net6.0 as an OOB: https://github.com/dotnet/runtime/blob/1f3c12be2192e1aad08271c0af4e23659c9532f8/src/libraries/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj#L3-L11

If there's a problem, my understanding is that I should adjust how we target netstandard2.0 with this package.


Of all our merged packages, System.Memory is one that looks very similar to this one: https://github.com/dotnet/maintenance-packages/blob/284e1dbb2222a6aeb02a28a02168eee55c27c37e/src/System.Memory/src/System.Memory.csproj#L12-L20

So this is what I initially choose for System.Runtime.CompilerServices.Unsafe:

```xml
<VersionPrefix>4.7.1</VersionPrefix>
<VersionPrefix Condition="'$(IsPackable)' == 'true'">4.8.0</VersionPrefix>
<AssemblyVersion Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">4.0.7.0</AssemblyVersion><!-- NO-INCREMENT: This version is frozen for .NET Standard and .NETCoreApp because the assembly ships inbox. -->
<AssemblyVersion Condition="'$(IsPackable)' == 'true' and $([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">4.0.8.0</AssemblyVersion>
<PackageValidationBaselineVersion>4.7.1</PackageValidationBaselineVersion>
<IsPlaceholderTargetFramework Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">true</IsPlaceholderTargetFramework>
```

But it fails saying that I am using the wrong AssemblyVersion:

```
Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error : API compatibility errors between 'lib/netstandard2.0/System.Runtime.CompilerServices.Unsafe.dll' (left) and 'lib/net462/System.Runtime.CompilerServices.Unsafe.dll' (right): [C:\Users\calope\source\repos\maintenance-packages\src\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj]
Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0003: lib/net462/System.Runtime.CompilerServices.Unsafe.dll assembly version '4.0.8.0' should be equal to or higher than lib/netstandard2.0/System.Runtime.CompilerServices.Unsafe.dll version '4.8.0.0'. [C:\Users\calope\source\repos\maintenance-packages\src\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj]
Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error : API breaking changes found. If those are intentional, the APICompat suppression file can be updated by rebuilding with '/p:ApiCompatGenerateSuppressionFile=true' [C:\Users\calope\source\repos\maintenance-packages\src\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj]
```

If the comment says:

```xml
<!-- NO-INCREMENT: This version is frozen for .NET Standard and .NETCoreApp because the assembly ships inbox. -->
```

Then the fix should probably be:

```xml
<AssemblyVersion Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">4.8.0.0</AssemblyVersion><!-- NO-INCREMENT: This version is frozen for .NET Standard and .NETCoreApp because the assembly ships inbox. -->
<AssemblyVersion Condition="'$(IsPackable)' == 'true' and $([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">4.8.1.0</AssemblyVersion>
```

And the ApiCompat failures are gone.

I see that we have more modern msbuild code in the net6.0 OOB package, so the next thing is to try to match what we use there instead of what we were using in netcoreapp3.1.

carlossanlop commented:
@ericstj

I am having trouble testing a locally generated version of a maintenance-packages nupkg as suggested by Viktor. This is what I did:

- I ran `dotnet pack src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj`
- It successfully generated a package named `System.Runtime.CompilerServices.Unsafe.4.8.0-dev.24376.1.nupkg`
- I extracted the contents of the nupkg into `C:/.tools/nuget/packages/System.Runtime.CompilerServices.Unsafe/4.8.0-dev.24376.1/`.
- Then in a brand new console app I added this:
```xml
 <ItemGroup>
  <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="4.8.0-dev.24376.1" />
</ItemGroup>
```
- But I get this error:
```
C:\Users\calope\source\repos\test\net9\tests.csproj : warning NU1603: tests depends on System.Runtime.CompilerServices.Unsafe (>= 4.8.0-dev.24376.1) but System.Runtime.CompilerServices.Unsafe 4.8.0-dev.24376.1 was not found. An approximate best match of System.Runtime.CompilerServices.Unsafe 5.0.0-preview.1.20120.5 was resolved.
```

Can you tell me what I'm doing wrong, or how you would do it?

ericstj commented:
> I extracted the contents of the nupkg into C:/.tools/nuget/packages/System.Runtime.CompilerServices.Unsafe/4.8.0-dev.24376.1/.

Don't manually put stuff in the package cache.  Instead you want to add a directory source to your nuget.config.  So for your test project, make sure the nuget config has:
```
  <packageSources>
    <add key="maintenance-packages-privates" value="C:\src\dotnet\maintenance-packages\artifacts\packages\Release\Shipping" />
  </packageSources>
```

carlossanlop commented:
Ok thanks, @ericstj. I don't see any catastrophic failures when using the local package in a test project targeting net6.0. I did this:

NuGet.config in the test project:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
  <packageSources>
    <clear />
    <add key="maintenance-packages-prerelease" value="C:\Users\calope\source\repos\maintenance-packages\artifacts\packages\Release\Shipping\" />
  </packageSources>
</configuration>
```

`dotnet add package System.Runtime.CompilerServices.Unsafe` added this to the csproj:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="4.8.0-dev.24379.1" />
  </ItemGroup>
</Project>
```
My Program.cs does this:
```cs
using System;
Console.WriteLine(typeof(System.Runtime.CompilerServices.Unsafe).Assembly.Location);
```
And the output is:
```
> dotnet run
C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.32\System.Runtime.CompilerServices.Unsafe.dll
```

Consuming that package is expected since 6.0 [publishes its own OOB](https://github.com/dotnet/runtime/blob/1f3c12be2192e1aad08271c0af4e23659c9532f8/src/libraries/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj#L10).

carlossanlop commented:
I'm getting this error now:

> C:\.tools\.nuget\packages\microsoft.dotnet.arcade.sdk\8.0.0-beta.24367.1\tools\SymStore.targets(64,5): error MSB3073: The command ""C:\.tools\.nuget\packages\microsoft.diasymreader.pdb2pdb\1.1.0-beta2-19575-01\tools\Pdb2Pdb.exe" "C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Runtime.CompilerServices.Unsafe\Release\netstandard2.0\System.Runtime.CompilerServices.Unsafe.dll" /out "C:\Users\calope\source\repos\maintenance-packages\artifacts\SymStore\Release\System.Runtime.CompilerServices.Unsafe\netstandard2.0\System.Runtime.CompilerServices.Unsafe.pdb" /srcsvrvar SRC_INDEX=public" exited with code 2. [C:\Users\calope\source\repos\maintenance-packages\src\System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj::TargetFramework=netstandard2.0]


If I run that command manually:

```
$ C:\.tools\.nuget\packages\microsoft.diasymreader.pdb2pdb\1.1.0-beta2-19575-01\tools\Pdb2Pdb.exe "C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Runtime.CompilerServices.Unsafe\Release\netstandard2.0\System.Runtime.CompilerServices.Unsafe.dll" /out "C:\Users\calope\source\repos\maintenance-packages\artifacts\SymStore\Release\System.Runtime.CompilerServices.Unsafe\netstandard2.0\System.Runtime.CompilerServices.Unsafe.pdb" /srcsvrvar SRC_INDEX=public
```

It says the pdb is not found or doesn't match:

> File 'C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Runtime.CompilerServices.Unsafe\Release\netstandard2.0\System.Runtime.CompilerServices.Unsafe.pdb' not found or doesn't match 'C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Runtime.CompilerServices.Unsafe\Release\netstandard2.0\System.Runtime.CompilerServices.Unsafe.dll'.

ericstj commented on src/System.Runtime.CompilerServices.Unsafe/ver/Private.Versioning.csproj:
```diff
@@ -0,0 +1,7 @@
+
+  <PropertyGroup>
+    <TargetFrameworks>$(NetFrameworkMinimum);netstandard2.0</TargetFrameworks>
+  </PropertyGroup>
```
We either need to add the version properties here, or share them in a props file.

ericstj commented on Directory.Build.props:
```diff
@@ -6,7 +6,6 @@
     <Copyright>$(CopyrightNetFoundation)</Copyright>
     <PackageLicenseExpression>MIT</PackageLicenseExpression>
     <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
-    <DebugType>embedded</DebugType>
```
Per our conversation - make sure we set DebugType=Portable and `IncludeSymbols` here.

carlossanlop replied:
Will do. And `DocFileItem` too.

ericstj commented on Directory.Build.targets:
```diff
@@ -6,6 +6,11 @@
 
+  <PropertyGroup>
+    <!-- SDK sets product to assembly but we want it to be our product name -->
+    <Product>Microsoft%AE .NET</Product>
```
Consider moving to props with other settings.  SDK won't overwrite it once set.

ericstj commented on src/System.Runtime.CompilerServices.Unsafe/ver/System.Runtime.CompilerServices.Unsafe.Versioning.csproj:
```diff
@@ -0,0 +1,7 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(NetFrameworkMinimum);netstandard2.0</TargetFrameworks>
```
Nit: I think you only need `netstandard2.0`

carlossanlop replied:
Hmm the ProjectReference is conidtioned to all except net7.0:
```xml
  <ItemGroup Condition="'$(IsPlaceholderTargetFramework)' != 'true'">
    <ProjectReference Include="..\ver\System.Runtime.CompilerServices.Unsafe.Versioning.csproj"
                      PrivateAssets="All"
                      ReferenceOutputAssembly="false"
                      OutputItemType="ILResourceReference" />
```
Should we change that condition to only import the project for ns2.0?

carlossanlop replied:
@ericstj No, I need NetFrameworkMinimum here. I built with with IsPackable=true, opened the DLLs for net462 using ILSpy, and with NetFrameworkMinimum included, the AssemblyVersion is correct (6.0.1.0), but it's incorrect if I remove NetFrameworkMinimum (6.0.0.0).

ericstj replied:
I would only expect this to impact *file version* assembly version should be determined by the AssemblyVersion properties in the ilproj.  I would expect file version to be the same for everything.

The reason I said you could remove `NetFrameworkMinimum` is that the build of the ilproj for `NetFrameworkMinimum` can just use the `netstandard2.0` copy of the Versioning project.

Also -- I thought we were going to give this a different name like "private.version" or something to avoid accidentally shipping it.

carlossanlop replied:
The name we choose for this versioning project is what shows up in the DLL Properties -> Details -> Product name. I thought we would prefer it to be named "System.Runtime.CompilerServices.Unsafe.Versioning" rather than seeing just "Version".

ericstj replied:
Oh, that's interesting.  Nice catch!  In that case let's keep your project name, but lets make the assembly named "System.Runtime.CompilerServices.Unsafe" so that we get the right value in that field of the version properties.

carlossanlop replied:
I'll try in a separate PR. If I recall correctly, using that same name caused build issues. Will let you know tomorrow.

carlossanlop replied:
@ericstj do you have a strong opinion on the name this project should have?

ericstj replied:
You can't have the same project name.  They fight over the output directory.  That's why I mentioned above to set `AssemblyName`
> In that case let's keep your project name, but lets make the assembly named "System.Runtime.CompilerServices.Unsafe"

Does this work?

carlossanlop replied:
Oooooooh! I didn't know there was a property for that.

ericstj replied:
https://learn.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-properties?view=vs-2022#list-of-common-properties-and-parameters

carlossanlop commented:
I have addressed the latest suggestions. I will squash now and then merge.

ViktorHofer commented on src/System.Runtime.CompilerServices.Unsafe/Directory.Build.props:
```diff
@@ -0,0 +1,10 @@
+<Project>
+
+  <Import Project="..\Directory.Build.props" />
```
The SDK team recommends to use the GetPathOfFileAbove function so that you don't need to resolve the path to the next outer D.B.props file manually or need an intermediate (empty) Directory.Build.props file:

```xml
<Import Project="$([MSBuild]::GetPathOfFileAbove(Directory.Build.props, $(MSBuildThisFileDirectory)..))" />
```

carlossanlop replied:
@ericstj Isn't this the one-liner you were talking about?

ericstj replied:
Yep, exactly.

ViktorHofer commented on src/Directory.Build.props:
This file should be deleted, it doesn't serve any purpose.

carlossanlop replied:
@ericstj and I talked about this and preferred to have "a directory in the path above" in case anything changed in the directory layout.

But if I change it to the one-liner you shared in the other comment, then yes, we can remove this.

ericstj replied:
Please use what Viktor shared
